pkg/ds/farm.go

```diff
@@ -0,0 +1,358 @@
+	if err != nil {
+		return "", false, util.Errorf("Error retrieving eligible nodes for daemon set: %v", err)
+	}
+
+	// If this daemon set has a node selector set to Everything, check the labels
+	// of other daemon sets
+	for _, child := range dsf.children {
+		everythingSelector := klabels.Everything().String()
+		if !child.ds.IsDisabled() && child.ds.PodID() == dsFields.PodID && child.ds.ID() != dsFields.ID {
+			// Naively check if both selectors are the Everything selector plus
+			// something else or if both selectors are the same
+			//
+			// It is possible that both selectors won't contend if they both have
+			// something like { az = zone_one , az != zone_one } and that this
+			// would still report it as a contention, this behavior will remain
+			// until it needs to be changed as catching useless selectors is a
+			// good idea too
+
+			// If either the child or the current daemon set has the everything selector
+			// then they contend
+			if dsFields.NodeSelector.String() == everythingSelector ||
+				child.ds.GetNodeSelector().String() == everythingSelector {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+
+			// If both daemon sets have the same selector, then they contend
+			if dsFields.NodeSelector.String() == child.ds.GetNodeSelector().String() {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+		}
+	}
+
+	for _, child := range dsf.children {
+		// Check child if it is not disabled and that if has the same PodID
+		if !child.ds.IsDisabled() && child.ds.PodID() == dsFields.PodID && child.ds.ID() != dsFields.ID {
+			// Check the child's eligibleNodes, then intersect it to see if there
+			// are any overlaps
+			//
+			// NOTE: This is naive, it does not account for new nodes, so any alerts
+			// we get will be caused by adding new nodes by both human, machine error,
+			// are starting up a daemon set farm where contention already exists
+			scheduledNodes, err := child.ds.EligibleNodes()
+			if err != nil {
+				return "", false, util.Errorf("Error getting scheduled nodes: %v", err)
+			}
+			intersectedNodes := types.NewNodeSet(eligibleNodes...).Intersection(types.NewNodeSet(scheduledNodes...))
+			if intersectedNodes.Len() > 0 {
+				dsf.raiseContentionAlert(child.ds.ID(), dsFields.ID)
+				return child.ds.ID(), true, nil
+			}
+		}
+	}
+
+	return "", false, nil
+}
+
+func (dsf *Farm) raiseContentionAlert(oldDSID ds_fields.ID, newDSID ds_fields.ID) {
+	if alertErr := dsf.alerter.Alert(alerting.AlertInfo{
+		Description: fmt.Sprintf("New ds '%v', contends with '%v'", newDSID, oldDSID),
+		IncidentKey: "preemptive_ds_contention",
+	}); alertErr != nil {
+		dsf.logger.WithError(alertErr).Errorln("Unable to deliver alert!")
+	}
+}
+
+// Creates a functioning daemon set that will watch and write to the pod tree
+func (dsf *Farm) spawnDaemonSet(dsFields *ds_fields.DaemonSet, quitCh <-chan struct{}) *childDS {
```

Comment:
Updated and removed the parameter to the function. Still making a new quit channel.